File: LogicAndTrick.WikiCodeParser/Tags/SpoilerTag.cs
using System;
using System.Collections.Generic;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Tags
{
    public class SpoilerTag : Tag
    {
        public SpoilerTag()
        {
            Token = "spoiler";
            Element = "span";
            ElementClass = "spoiler";
            MainOption = "text";
            Options = new[] {"text"};
            AllOptionsInMain = true;
        }

        public override INode FormatResult(Parser parser, ParseData data, State state, string scope, Dictionary<string, string> options, string text)
        {
            var visibleText = "Spoiler";
            if (options.ContainsKey("text") && options["text"].Length > 0) visibleText = options["text"];

            var before = "<" + Element;
            if (ElementClass != null) before += " class=\"" + ElementClass + '"';
            before += $" title=\"{visibleText}\">";
            var after = "</" + Element + '>';
            return new HtmlNode(before, new SpoilerNode(visibleText, parser.ParseTags(data, text, scope, IsBlock ? "block" : "inline")), after);
        }

        private class SpoilerNode : INode
        {
            private string VisibleText { get; }
            private INode SpoilerContent { get; set; }

            public SpoilerNode(string visibleText, INode spoilerContent)
            {
                VisibleText = visibleText;
                SpoilerContent = spoilerContent;
            }

            public string ToHtml()
            {
                return SpoilerContent.ToHtml();
            }

            public string ToPlainText()
            {
                return $"[{VisibleText}](spoiler text)";
            }

            public IList<INode> GetChildren()
            {
                return new[] {SpoilerContent};
            }

            public void ReplaceChild(int i, INode node)
            {
                if (i != 0) throw new ArgumentOutOfRangeException();
                SpoilerContent = node;
            }
        }
    }
}

File: javascript/src/Tags/SpoilerTag.ts
import { Parser } from '..';
import { HtmlNode } from '../Nodes/HtmlNode';
import { INode } from '../Nodes/INode';
import { ParseData } from '../ParseData';
import { State } from '../State';
import { Tag } from './Tag';

class SpoilerNode implements INode {
    private VisibleText: string;
    private SpoilerContent: INode;

    constructor(visibleText: string, spoilerContent: INode) {
        this.VisibleText = visibleText;
        this.SpoilerContent = spoilerContent;
    }

    ToHtml(): string {
        return this.SpoilerContent.ToHtml();
    }

    ToPlainText(): string {
        return `[${this.VisibleText}](spoiler text)`;
    }

    GetChildren(): INode[] {
        return [this.SpoilerContent];
    }

    ReplaceChild(i: number, node: INode): void {
        if (i != 0) throw new Error('Argument out of range');
        this.SpoilerContent = node;
    }

    HasContent(): boolean {
        return true;
    }
}

export class SpoilerTag extends Tag {
    constructor() {
        super();
        this.Token = 'spoiler';
        this.Element = 'span';
        this.ElementClass = 'spoiler';
        this.MainOption = 'text';
        this.Options = ['text'];
        this.AllOptionsInMain = true;
    }

    public FormatResult(parser: Parser, data: ParseData, state: State, scope: string, options: Record<string, string>, text: string): INode {
        let visibleText = 'Spoiler';
        if (options['text'] && options['text'].length > 0) visibleText = options['text'];

        let before = `<${this.Element}`;
        if (this.ElementClass != null) before += ' class="' + this.ElementClass + '"';
        before += ` title="${visibleText}">`;
        const after = `</${this.Element}>`;
        return new HtmlNode(before, new SpoilerNode(visibleText, parser.ParseTags(data, text, scope, this.TagContext())), after);
    }
}
